// @imagekit/nodejs/resources/saved-extensions.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.SavedExtensions = void 0;
const resource_1 = require("../core/resource.js");
const headers_1 = require("../internal/headers.js");
const path_1 = require("../internal/utils/path.js");
class SavedExtensions extends resource_1.APIResource {
    /**
     * This API creates a new saved extension. Saved extensions allow you to save
     * complex extension configurations (like AI tasks) and reuse them by referencing
     * the ID in upload or update file APIs.
     *
     * **Saved extension limit** \
     * You can create a maximum of 100 saved extensions per account.
     *
     * @example
     * ```ts
     * const savedExtension = await client.savedExtensions.create({
     *   config: {
     *     name: 'ai-tasks',
     *     tasks: [
     *       {
     *         instruction: 'What types of clothing items are visible in this image?',
     *         type: 'select_tags',
     *         vocabulary: ['shirt', 'dress', 'pants', 'jacket', 'shoes'],
     *         max_selections: 3,
     *       },
     *     ],
     *   },
     *   description: 'Automatically categorizes clothing items in fashion images',
     *   name: 'Fashion Item Categorization',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/v1/saved-extensions', { body, ...options });
    }
    /**
     * This API updates an existing saved extension. You can update the name,
     * description, or config.
     *
     * @example
     * ```ts
     * const savedExtension = await client.savedExtensions.update(
     *   'id',
     * );
     * ```
     */
    update(id, body, options) {
        return this._client.patch((0, path_1.path) `/v1/saved-extensions/${id}`, { body, ...options });
    }
    /**
     * This API returns an array of all saved extensions for your account. Saved
     * extensions allow you to save complex extension configurations and reuse them by
     * referencing them by ID in upload or update file APIs.
     *
     * @example
     * ```ts
     * const savedExtensions = await client.savedExtensions.list();
     * ```
     */
    list(options) {
        return this._client.get('/v1/saved-extensions', options);
    }
    /**
     * This API deletes a saved extension permanently.
     *
     * @example
     * ```ts
     * await client.savedExtensions.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/v1/saved-extensions/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * This API returns details of a specific saved extension by ID.
     *
     * @example
     * ```ts
     * const savedExtension = await client.savedExtensions.get(
     *   'id',
     * );
     * ```
     */
    get(id, options) {
        return this._client.get((0, path_1.path) `/v1/saved-extensions/${id}`, options);
    }
}
exports.SavedExtensions = SavedExtensions;
